src/webgpu/shader/execution/expression/constructor/non_zero.spec.ts

```diff
@@ -0,0 +1,419 @@
+      [elementType],
+      vectorType,
+      t.params,
+      [{ input: [elementType.create(value)], expected: vectorType.create(value) }]
+    );
+  });
+
+g.test('vector_elements')
+  .specURL('https://www.w3.org/TR/WGSL/#value-constructor-builtin-function')
+  .desc(`Test that a vector constructed from element values produces the expected value`)
+  .params(u =>
+    u
+      .combine('inputSource', allInputSources)
+      .combine('type', ['bool', 'i32', 'u32', 'f32', 'f16'] as const)
+      .combine('width', [2, 3, 4] as const)
+      .combine('infer_type', [false, true] as const)
+  )
+  .beforeAllSubcases(t => {
+    if (t.params.type === 'f16') {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(async t => {
+    const elementType = Type[t.params.type];
+    const vectorType = Type.vec(t.params.width, elementType);
+    const elements: number[] = [];
+    const fn = t.params.infer_type ? `vec${t.params.width}` : `${vectorType}`;
+    for (let i = 0; i < t.params.width; i++) {
+      if (t.params.type === 'bool') {
+        elements.push(i & 1);
+      } else {
+        elements.push((i + 1) * 10);
+      }
+    }
+
+    await run(
+      t,
+      basicExpressionBuilder(ops => `${fn}(${ops.join(', ')})`),
+      elements.map(e => elementType),
+      vectorType,
+      t.params,
+      [
+        {
+          input: elements.map(v => elementType.create(v)),
+          expected: vectorType.create(elements),
+        },
+      ]
+    );
+  });
+
+g.test('vector_mix')
+  .specURL('https://www.w3.org/TR/WGSL/#value-constructor-builtin-function')
+  .desc(
+    `Test that a vector constructed from a mix of element values and sub-vectors produces the expected value`
+  )
+  .params(u =>
+    u
+      .combine('inputSource', allInputSources)
+      .combine('type', ['bool', 'i32', 'u32', 'f32', 'f16'] as const)
+      .combine('signature', [
+        '2s', //   [vec2,   scalar]
+        's2', //   [scalar, vec2]
+        '2ss', //  [vec2,   scalar,   scalar]
+        's2s', //  [scalar, vec2,     scalar]
+        'ss2', //  [scalar, scalar,   vec2  ]
+        '22', //   [vec2,   vec2]
+        '3s', //   [vec3,   scalar]
+        's3', //   [scalar, vec3]
+      ] as const)
+      .combine('infer_type', [false, true] as const)
+  )
+  .beforeAllSubcases(t => {
+    if (t.params.type === 'f16') {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(async t => {
+    const elementType = Type[t.params.type];
+    let width = 0;
+    const elementValue = (i: number) => (t.params.type === 'bool' ? i & 1 : (i + 1) * 10);
+    const elements: number[] = [];
+    const nextValue = () => {
+      const value = elementValue(width++);
+      elements.push(value);
+      return elementType.create(value);
+    };
+    const args: Value[] = [];
+    for (const c of t.params.signature) {
+      switch (c) {
+        case '2':
+          args.push(vec2(nextValue(), nextValue()));
```

Comment:
relies on parameter evaluation order.
 ECMAscript specifies left-to-right.

LGTM